### PR TITLE
Improve state management and triggers #529 #530 #519 #520 #510 #507

### PR DESCRIPTION
**Pull request by lbradstreet**
Still required:
Trigger state scoping e.g. there's currently a trigger state maintained for each window e.g. every 5 segments for a window will cause the window to fire, but it may be the case that you want to fire every 5 segments for all the windows, or possibly every 5 segments for any grouping key.

Closes: 
#529 #530 #519 #520 #510 #507

Still left to be done:
Parameterizable notifications #488

### CODE REVIEW

**Comment by lbradstreet** [comment]
I just realised that https://github.com/onyx-platform/onyx/blob/0.8.x/doc/user-guide/triggers.md is out of date now. Fix this up before we merge.

File: src/onyx/information_model.cljc [hunk 1]
```diff
@@ -380,18 +380,45 @@
...
                                :added "0.8.0"}
-            :aggregation/fn {:doc "Fn (state, window, segment) to generate a serializable state machine update."
-                             :type :function
-                             :optional? false
-                             :added "0.8.0"}
-            :aggregation/apply-state-update {:doc "Fn (state, entry) to apply state machine update entry to a state."
+            :aggregation/create-state-update {:doc "Fn (window, state, segment) to generate a serializable state machine update."
+                                              :type :function
+                                              :optional? false
+                                              :added "0.8.0"}
```
Comment (MichaelDrogalis) [inline src/onyx/information_model.cljc]: Let's bump the version to 0.9.0.

File: src/onyx/information_model.cljc [hunk 2]
```diff
@@ -380,18 +380,45 @@
...
                                                :optional? true
                                                :added "0.8.0"}}}
+
+   :state-refinement
+   {:summary "Onyx provides the ability to perform state refinements after triggers fired."
+    :link nil
+    :model {:refinement/create-state-update {:doc "Fn (trigger, state, state-event) to generate a serializable state machine update."
+                                              :type :function
+                                              :optional? false
+                                              :added "0.0.0"}
```
Comment (MichaelDrogalis) [inline src/onyx/information_model.cljc]: Switch version.

File: src/onyx/information_model.cljc [hunk 3]
```diff
@@ -380,18 +380,45 @@
...
+            :refinement/apply-state-update {:doc "Fn (trigger, state, entry) to apply the refinement state machine update entry to a state."
+                                             :type :function
+                                             :optional? false
+                                             :added "0.8.0"}}}
```
Comment (MichaelDrogalis) [inline src/onyx/information_model.cljc]: Version.

File: src/onyx/information_model.cljc [hunk 4]
```diff
@@ -658,6 +696,13 @@ You can use lifecycles to supply any stateful connections necessary to sync your
              :default []
              :added "0.8.9"}
 
+            :onyx.peer/trigger-timer-resolution
+            {:doc "The resolution of the timer firing state-events that are not caused by segments arriving."
+             :type :integer
+             :optional? true
+             :default 100
```
Comment (MichaelDrogalis) [inline src/onyx/information_model.cljc]: Let's add a `:units` key here. Milliseconds, I presume?

File: src/onyx/peer/task_lifecycle.clj [hunk 5]
```diff
@@ -368,6 +323,16 @@
             (extensions/write-chunk log :exception e job-id)
             (>!! outbox-ch entry))))))
 
+(s/defn assign-windows :- Event
```
Comment (MichaelDrogalis) [inline src/onyx/peer/task_lifecycle.clj]: This area got so much cleaner. Nice work.

File: src/onyx/information_model.cljc (hunk 1 above)
Comment (lbradstreet) [inline src/onyx/information_model.cljc]: These were present in 0.8.0. You're right about the others though, I'm not sure how I missed them.

File: src/onyx/triggers.clj [hunk 6]
```diff
@@ -0,0 +1,144 @@
+(ns onyx.triggers
```
Comment (MichaelDrogalis) [inline src/onyx/triggers.clj]: Can we move this file back down into the triggers directory?

**Comment by MichaelDrogalis** [comment]
Looks good on first pass, but definitely need to talk about it during a review when we're both around. Clear to merge. :+1:

**Pull request merged**